Sources/Articles/Documentation.docc/2025/gha-faster-ci/gha-faster-ci.md

```diff
@@ -0,0 +1,547 @@
+# Faster CI in GitHub Actions
+
+Swift has a lot of power points, but performance of the build system is not one of them. It can easily take anything between 10 to 30 minutes for a typical Swift CI to run depending on how big the project is, what build configuration you're using, how beefy your build machines are, and other factors.
+
+By optimizing your CI runtime you'll save precious developer time as well as either paying less for CI or consuming less of your GitHub Actions free quota.
+
+In this article, you'll walk through optimizing [Vapor's Penny Bot](https://github.com/vapor/penny-bot) CI times to go from 10 minutes in tests and 14.5 minutes in deployments, down to less than 4 minutes in tests, and 3 minutes in deployments. The bigger your project is, the bigger the gap will be.
+
+To follow along, you should checkout the same code that this article is based on: `git checkout swift-on-server-article`.
+
+## The Problem
+
+In [GitHub Actions](https://docs.github.com/actions), a usual CI file to run tests of a Swift project can look like this:
+
+```yaml
+name: tests
+on:
+  pull_request: { types: [opened, reopened, synchronize, ready_for_review] }
+  push: { branches: [main] }
+
+jobs:
+  unit-tests:
+    runs-on: ubuntu-latest
+    container: swift:6.0-noble
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Run unit tests
+        run: swift test --enable-code-coverage
+
+      # Process the code coverage report, etc...
+```
+
+For deployments you'd use a CI file like below in combination with [Vapor](https://github.com/vapor/template/blob/main/Dockerfile) or [Hummingbird](https://github.com/hummingbird-project/template/blob/main/Dockerfile) template's Dockerfile.
+
+This enables you to deploy your apps to cloud services that consume Docker images, such as AWS ECS or DigitalOcean App Platform.
+
+```yaml
+name: deploy
+on:
+  push: { branches: [main] }
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Build image
+        run: docker build --network=host -t app:latest .
+
+      # Push the image to a Docker container registry and deploy the app
+```
+
+These CIs usually take 10 minutes in tests and 14.5 minutes for deployments in [Penny](https://github.com/vapor/penny-bot).
+
+That's too much time wasted waiting. How can you improve these CI times?
+
+![Tests CI Initial State](tests-ci-initial.png)
+![Deployment CI Initial State](deploy-ci-initial.png)
+
+## Use Caching
```

Comment:
How about adding a small section here to run tests using the `--parallel` flag? 
Some users are not aware of this I suppose, but it'll certainly speed up unit tests.

Reply:
hmm i think a short mention should be good enough because the problem is usually if you just enable --parallel you'll have a bunch of failures due to shared resources, so it won't be like a "magic" flag that i can write about. More like a suggestion for users to think about and pursue 🤔

Reply:
Yes, I was thinking about just a short mention, completely valid points about shared resource issues. 😉

Reply:
Ended up with a whole new section: https://github.com/swift-on-server/articles/pull/23/commits/2f1e29e2478f56d49dca4bbd060444e3ad48ba8f